tests/zkevm/test_worst_memory.py

```diff
@@ -49,37 +50,71 @@ class CallDataOrigin:
         pytest.param(1024 * 1024, id="1MiB"),
     ],
 )
-@pytest.mark.parametrize("fixed_dst", [True, False])
+@pytest.mark.parametrize(
+    "fixed_src_dst",
+    [
+        True,
+        False,
+    ],
+)
+@pytest.mark.parametrize(
+    "non_zero_data",
+    [
+        True,
+        False,
+    ],
+)
 def test_worst_calldatacopy(
     state_test: StateTestFiller,
     pre: Alloc,
     fork: Fork,
     origin: CallDataOrigin,
     size: int,
-    fixed_dst: bool,
+    fixed_src_dst: bool,
+    non_zero_data: bool,
 ):
     """Test running a block filled with CALLDATACOPY executions."""
     env = Environment()
 
+    if size == 0 and non_zero_data:
+        pytest.skip("Non-zero data with size 0 is not applicable.")
+
     # We create the contract that will be doing the CALLDATACOPY multiple times.
-    dst = 0 if fixed_dst else Op.MOD(Op.GAS, 7)
-    attack_block = Op.CALLDATACOPY(dst, Op.PUSH0, Op.CALLDATASIZE)
-    code = code_loop_precompile_call(Bytecode(), attack_block, fork)
+    #
+    # If `non_zero_data` is True, we leverage CALLDATASIZE for the copy length. Otherwise, since we
+    # don't send zero data explicitly via calldata, PUSH the target size and use DUP1 to copy it.
+    prefix = Bytecode() if non_zero_data or size == 0 else Op.PUSH32(size)
+    src_dst = 0 if fixed_src_dst else Op.MOD(Op.GAS, 7)
+    attack_block = Op.CALLDATACOPY(
+        src_dst, src_dst, Op.CALLDATASIZE if non_zero_data or size == 0 else Op.DUP1
+    )
+    code = code_loop_precompile_call(prefix, attack_block, fork)
```

Comment:
Two things:
- `non_zero_data` now means being more careful about assuming `CALLDATASIZE` is the right "size" value.
- For `size==0` is a special^2 case (since even for `not non_zero_data` it doesn't make sense to do `Op.PUSH32(size)` thus we can leverage `CALLDATASIZE` (which is 0) -- hope that isn't confusing and maybe I'm squeezing too much ha)

Reply:
Hmm this looks confusing, because there is now a stack imbalance at the prefix. That either initializes the stack with 0 or 1 items depending on the parameters. AH! Wait, I see now, for `non_zero_data or size == 0` items we use Op.CALLDATASIZE which thus does not have the Op.DUP1 stack requirement of having an item.

For the 0 case for clarity could also use PUSH0 for the `size==0` case. 

We should take a look if there is a builtin method to pick the right PUSH here (since PUSH32 seems overkill). If you would just use `size` instead of `Op.PUSH32(size)` would it then not pick the "smallest" PUSH?

Reply:
Changed to `PUSH24` under the logic of a `size` of 10MiB (10x than the current max one, just to be safe):
![image](https://github.com/user-attachments/assets/ff4c293e-860a-4923-bccc-b43f56490ee3)

Reply:
Ah no, I meant "just use the size" as in remove the `Op.PUSH32(size)` and just keep `size`. I think the test formatter will pick the appropriate PUSH value.

For 10 MiB you can actually store it in PUSH3 (24 bits) :wink:

Reply:
@jochem-brouwer , you mean doing this?
`prefix = Bytecode() if non_zero_data or size == 0 else size`
I don't think that would work.

I think ideally we should have a way to use [this private func](https://github.com/ethereum/execution-spec-tests/blob/jsign-mem-opcodes/src/ethereum_test_vm/opcode.py#L33).